WalletWasabi.Fluent/ViewModels/Wallets/Send/TransactionPreviewViewModel.cs

```diff
@@ -296,23 +296,48 @@ private async Task<bool> InitialiseTransactionAsync()
 
 			var result = await TryAdjustTransactionFeeAsync(percentage);
 
-			return result ? await BuildTransactionAsync(reason) : null;
+			return result ? await BuildTransactionAsync() : null;
 		}
 		catch (TransactionFeeOverpaymentException ex)
 		{
 			var result = await TryAdjustTransactionFeeAsync(ex.PercentageOfOverpayment);
 
-			return result ? await BuildTransactionAsync(reason) : null;
+			return result ? await BuildTransactionAsync() : null;
 		}
 		catch (InsufficientBalanceException ex)
 		{
 			var failedTransactionFee = ex.Minimum - _info.Amount;
 			var maxPossibleFeeWithSelectedCoins = ex.Actual - _info.Amount;
 			var differenceOfFeePercentage = maxPossibleFeeWithSelectedCoins == Money.Zero ? 0M : (decimal)failedTransactionFee.Satoshi / maxPossibleFeeWithSelectedCoins.Satoshi * 100;
 
-			var result = await TryHandleInsufficientBalanceCaseAsync(differenceOfFeePercentage, ex.Minimum, reason);
+			var canSelectMoreCoins = _wallet.Coins.Any(coin => !_info.Coins.Contains(coin));
+			var isMaxFeeRateFound = TransactionFeeHelper.TryGetMaximumPossibleFeeRate(differenceOfFeePercentage, _wallet, _info.FeeRate, out var maximumPossibleFeeRate);
 
-			return result ? await BuildTransactionAsync(reason) : null;
+			if (canSelectMoreCoins)
+			{
+				var selectPocketsDialog =
+					await NavigateDialogAsync(new PrivacyControlViewModel(_wallet, _info, usedCoins: Transaction?.SpentCoins, isSilent: true));
+
+				if (selectPocketsDialog.Result is { } newCoins)
+				{
+					_info.Coins = newCoins;
+					return await BuildTransactionAsync();
+				}
+			}
+			else if (isMaxFeeRateFound)
+			{
+				_info.MaximumPossibleFeeRate = maximumPossibleFeeRate;
+				_info.FeeRate = maximumPossibleFeeRate;
+				_info.ConfirmationTimeSpan = TransactionFeeHelper.CalculateConfirmationTime(maximumPossibleFeeRate, _wallet);
+				return await BuildTransactionAsync();
```

Comment:
As I mentioned before, it would be better to pass the data which is needed to build the transaction as call parameters instead of modifying the members and retrying. Always favor immutability (when it's possible).

Reply:
I took a note about this, will take a look. Thanks!